Source/WebGPU/WebGPU/BindGroup.mm

```diff
@@ -77,34 +110,39 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
             if (bufferIsPresent + samplerIsPresent + textureViewIsPresent != 1)
                 return BindGroup::createInvalid(*this);
 
-            auto stage = bindGroupLayout.stageForBinding(entry.binding) / 2;
-            auto renderStage = [] (uint32_t inputStage) {
-                return (MTLRenderStages)(inputStage * 2);
-            };
-            static_assert(!(WGPUShaderStage_Vertex / 2) && WGPUShaderStage_Fragment / 2 == 1 && WGPUShaderStage_Compute / 2 == 2, "Unexpected shader stage constants");
-            static_assert((int)WGPUShaderStage_Vertex == (int)MTLRenderStageVertex && (int)WGPUShaderStage_Fragment == (int)MTLRenderStageFragment, "Expect identical mapping of vertex and fragment stages");
-            const size_t offset = bindingForStage[stage] * sizeof(float*);
-            ++bindingForStage[stage];
-
-            if (bufferIsPresent) {
-                id<MTLBuffer> buffer = WebGPU::fromAPI(entry.buffer).buffer();
-                ASSERT(sizeof(float*) == sizeof(buffer.gpuAddress));
-                *(float**)(argumentBufferContents[stage] + offset) = (float*)buffer.gpuAddress;
-
-            } else if (samplerIsPresent) {
-                id<MTLSamplerState> sampler = WebGPU::fromAPI(entry.sampler).samplerState();
-                ASSERT(sizeof(float*) == sizeof(sampler.gpuResourceID));
-                *(MTLResourceID*)(argumentBufferContents[stage] + offset) = sampler.gpuResourceID;
-
-            } else if (textureViewIsPresent) {
-                id<MTLTexture> texture = WebGPU::fromAPI(entry.textureView).texture();
-                ASSERT(sizeof(float*) == sizeof(texture.gpuResourceID));
-                *(MTLResourceID*)(argumentBufferContents[stage] + offset) = texture.gpuResourceID;
-                resources.append({ texture, MTLResourceUsageRead, renderStage(stage) });
+            auto stages = bindGroupLayout ? bindGroupLayout->stagesForBinding(entry.binding) : (WGPUShaderStage_Vertex | WGPUShaderStage_Fragment | WGPUShaderStage_Compute);
+            for (size_t currentStage = 0; currentStage < shaderStageLength; ++currentStage) {
+                WGPUShaderStage renderStage = shaderStage[currentStage];
+                if (!(stages & renderStage))
+                    continue;
+
+                auto& offset = bindingOffsetForStage[currentStage];
+                ASSERT(argumentBufferContents[currentStage]);
+
+                if (bufferIsPresent) {
+                    id<MTLBuffer> buffer = WebGPU::fromAPI(entry.buffer).buffer();
+                    if (entry.offset > buffer.length)
+                        return BindGroup::createInvalid(*this);
+
+                    ASSERT(sizeof(float*) == sizeof(buffer.gpuAddress));
```

Comment:
Ditto about using MTLArgumentEncoder instead of gpuAddress. Might as well do it once in a cross-version way instead of having to do it twice

Reply:
I agree, I filed https://bugs.webkit.org/show_bug.cgi?id=250123 which I will get to soon, I did it this way since it seemed simpler at the time